Convert RW raw pointers to smart pointers

joaogvcarneiro opened the pull request:
* **Tickets addressed:** #1065
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
<!-- What approach was taken to satisfy the ticket being addressed? What should reviewers be aware of? How are your
commits organized? -->
A while back the RW effectors were upgraded such that the RW configuration data could be changed on the fly in python.  This was doing by passing a raw pointer of the python-created structure to the C code.  This also mean the user had to ensure this python configuration structure remained in memory. With the thruster effector upgrade the use of smart-pointers was explored successfully.  This branch upgrades the RW effectors to now also use smart-pointers. 

The first commit changes raw pointers to shared pointers in the ``.h`` and ``.cpp`` files. The second commit makes sure SWIG wraps the structures as shared pointers and not raw pointers. The third commit fixes errors in Python where raw pointers were used. The last commit adds release notes.

## Verification
<!-- How were the changes validated? Were any automated tests added, updated, removed, or re-baselined? If you didn't 
add or update any tests justify this choice. -->
RW unit tests were converted accordingly. All tests should be passing.

## Documentation
<!-- What documentation was invalidated by these changes? Which artifacts should reviewers check for accuracy and 
completeness? -->
N/A.

## Future work
<!-- What next steps can we anticipate from here, if any? -->
N/A.

joaogvcarneiro commented on src/utilities/simIncludeRW.py:
```diff
@@ -26,6 +26,7 @@
 import numpy
 from Basilisk.architecture import messaging
 from Basilisk.utilities import macros
+from Basilisk.simulation import reactionWheelStateEffector
```
I'd also prefer it if it were done like ``THRSimConfigPayload``. However, I couldn't get it to work because ``RWConfigMsgPayload`` is defined inside ``messaging``, whereas ``THRSimConfigPayload`` is defined inside ``dynamics``. From my understanding, message wrapping is done automatically for all messages, so you can't add a new .i file for a particular message; otherwise, there's duplicate SWIG code. We could maybe move the current message definition to reside inside ``dynamics``, and in that case, it would be easy to apply the same logic as before.

joaogvcarneiro replied:
That's a valid question. It's currently part of the C++ messages.

joaogvcarneiro commented on src/simulation/dynamics/reactionWheels/reactionWheelStateEffector.i:
```diff
@@ -38,6 +39,11 @@ from Basilisk.architecture.swig_common_model import *
 %include "reactionWheelStateEffector.h"
 %include "architecture/utilities/macroDefinitions.h"
 
+%shared_ptr(RWConfigMsgPayload)
```
I'll go ahead and resolve this because I assume this comment will be addressed alongside the first comment.

schaubh commented on src/simulation/thermal/motorThermal/_UnitTest/test_motorThermal.py:
```diff
@@ -221,7 +221,6 @@ def motorThermalTest(show_plots, accuracy):
     # change the test conditions
     thermalModel.currentTemperature = 20
     thermalModel.ambientTemperature = 0
-    RW.useRWfriction = True
```
Mm, yes, `useRWfriction` is an argument for the `create()` method and does nothing here.  Thus, write now the same conditions are run twice.  We should just delete the test run that had `RW.varUseRWfriction = True`?

joaogvcarneiro replied:
Sure, I'll delete that test. Again, this should be refactored, but that's not a priority. I'll create an issue to make sure we don't miss that.

schaubh reviewed (approved):
Thanks for these edits.  This is a good clean up of RW and changing the config structure to be a regular *.h file and not a message.

Need to address a minor pre-commit issue, otherwise good to go.

joaogvcarneiro commented:
> Thanks for these edits. This is a good clean up of RW and changing the config structure to be a regular *.h file and not a message.
> 
> Need to address a minor pre-commit issue, otherwise good to go.

I made the changes to pre-commit, but also refactored the rwStateEffector class. I added those commits at the end.

juan-g-bonilla commented:
Great PR. Always a fan of using smart ptr when possible. Also nice that you added those extra clean ups with the for loops.

LGTM.

schaubh commented on src/simulation/thermal/motorThermal/_UnitTest/test_motorThermal.py:
```diff
@@ -218,16 +218,6 @@ def motorThermalTest(show_plots, accuracy):
     unitTestSim.ExecuteSimulation()
     numTests += 1
 
-    # change the test conditions
```
I don't see you testing the ability to change the RW data on the fly?  We should have a unit test that verifies this behavior works.

schaubh commented on docs/source/Support/bskReleaseNotes.rst:
```diff
@@ -61,6 +61,7 @@ Version |release|
 - Support calling ``unsubscribe`` on input messages.
 - Fixed an issue where the :ref:`forceTorqueThrForceMapping` module's Reset() function did not zero all thruster settings correctly.
 - Updated ``canary`` workflow to run on all pull requests to the develop branch, providing feedback on compatibility with latest dependencies.
+- Converted RW data structures in ``reactionWheelStateEffector`` to shared pointers instead of raw pointers.
```
add a bullet to the known issues RST file that the way to create the RW configuration has moved aware from `messages`. If the user is using the RW factory class no code changes are needed.  If they are creating this on their own then then show them what needs to be changed.

schaubh reviewed (changes requested):
- needs rebasing as well.
- needs a unit test to check the update functionality
- needs statement in known issues RST file to address the changed way of creating a RW configuration structure.
